src/dbnode/storage/bootstrap/bootstrapper/uninitialized/source.go

```diff
@@ -0,0 +1,156 @@
+func newUninitializedSource(opts Options) bootstrap.Source {
+	return &uninitializedSource{
+		opts: opts,
+	}
+}
+
+func (s *uninitializedSource) Can(strategy bootstrap.Strategy) bool {
+	switch strategy {
+	case bootstrap.BootstrapSequential:
+		return true
+	}
+
+	return false
+}
+
+func (s *uninitializedSource) AvailableData(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	return s.availability(ns, shardsTimeRanges, runOpts)
+}
+
+func (s *uninitializedSource) AvailableIndex(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	return s.availability(ns, shardsTimeRanges, runOpts)
+}
+
+func (s *uninitializedSource) availability(
+	ns namespace.Metadata,
+	shardsTimeRanges result.ShardTimeRanges,
+	runOpts bootstrap.RunOptions,
+) result.ShardTimeRanges {
+	var (
+		topoState                = runOpts.InitialTopologyState()
+		availableShardTimeRanges = result.ShardTimeRanges{}
+	)
+
+	for shardIDUint := range shardsTimeRanges {
+		shardID := topology.ShardID(shardIDUint)
+		hostShardStates, ok := topoState.ShardStates[shardID]
+		if !ok {
+			// This shard was not part of the topology when the bootstrapping
+			// process began.
+			continue
+		}
+
+		// The basic idea for the algorithm is that on a shard-by-shard basis we
+		// need to determine if the namespace is "new" in the sense that it has
+		// never been completely initialized (reached a state where all the hosts
+		// in the topology are "available").
+		// In order to determine this, we simply count the number of hosts in the
+		// "initializing" state. If this number is larger than zero, than the
+		// namespace is "new".
+		// The one exception to this case is when we perform topology changes and
+		// we end up with one extra node that is initializing which should be offset
+		// by the corresponding node that is leaving. I.E if numInitializing > 0
+		// BUT numLeaving >= numInitializing then it is still not a new namespace.
+		// See the TestUnitializedSourceAvailableDataAndAvailableIndex test for more details.
+		var (
+			numInitializing = 0
+			numLeaving      = 0
+		)
+		for _, hostState := range hostShardStates {
+			shardState := hostState.ShardState
+			switch shardState {
+			case shard.Initializing:
+				numInitializing++
+			case shard.Unknown:
+				// TODO: Right now we ignore unknown shards (which biases us towards
+				// failing the bootstrap). Once this interface supports returning errors,
+				// we should return an error in this case because the cluster is in a bad
+				// state and the operator should make a decision on how they want to proceed.
+			case shard.Leaving:
+				numLeaving++
+			case shard.Available:
+			default:
+				// TODO(rartoul): Make this a hard error once we refactor the interface to support
+				// returning errors.
+				panic(fmt.Sprintf("unknown shard state: %v", shardState))
```

Comment:
As per last comment: 
> Can we prioritize and make these `AvailableXyz(...)` methods return an error?  It would be really catastrophic if all the nodes panic when something does something stupid and modifies topology data in etcd in an unexpected way.